Add Timeout Feature to Email Notification Template

beeankha opened the pull request:
##### SUMMARY
Related to Issue https://github.com/ansible/awx/issues/2080

##### ISSUE TYPE

 - Feature Pull Request


##### COMPONENT NAME
 - API
 - UI

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 3.0.0
```


##### ADDITIONAL INFORMATION
Adding a timeout feature to email notifications using information found in Django base class found here: https://github.com/django/django/blob/stable/1.11.x/django/core/mail/backends/smtp.py

Adding a timeout scrollable input area in the UI.

beeankha committed:
Change email notification success/fail messages and add a timeout feature

jakemcdermott reviewed (dismissed):
The code changes look good to me - Is this still WIP? 

Before merging you'll also want to give this a run with an email template that won't succeed just to confirm that providing the timeout in the notification config has the expected effect.

beeankha commented:
Thank you for approving this, @jakemcdermott !   It is indeed still a WIP, testing the back end part of this to make sure that the timeout part of the notification is getting activated as needed.

beeankha commented:
Below are the things I've tried so far to debug/figure out why failing notifications time out at 6 seconds no matter what.

**Logging**

Added logging to `email_backend.py`:

![emailbackend_logger](https://user-images.githubusercontent.com/28930622/53505067-20620e00-3a81-11e9-9b1a-03f5902f37cb.png)


**Remote Debugging**

Set a breakpoint with `sdb` in `email_backend.py`:
 
![emailbackend_sdbtrace](https://user-images.githubusercontent.com/28930622/53505174-58695100-3a81-11e9-8aa0-409cbd81dcb8.png)

Added an `sdb` breakpoint from inside of the AWX container to `smtp.py`: 

![djangobackendsdbtrace](https://user-images.githubusercontent.com/28930622/53505258-8484d200-3a81-11e9-91a5-387503dfa585.png)


Nothing obvious is showing up yet, continuing to test this timeout feature...

kdelee commented on awx/ui/client/src/notifications/notificationTemplates.form.js:
```diff
@@ -128,10 +128,25 @@ export default ['i18n', function(i18n) {
+                integer: true,
+                spinner: true,
+                min: 0,
+                default: 0,
```
should the default timeout be 0? Seems like very little time

beeankha replied:
It seems that a timeout value of 0 = "no timeout".  Still testing this, however.

beeankha commented:
**Current Work**

- Testing to see what the min/max/default times should be
- Troubleshooting to see why there is an "in progress" fail message when the timeout is set to nothing (even when taking the `default: 0` out)
- Since this function is in there, does it now need to be required?  When left blank, it's failing every time.

**Progress That's Been Made**

There used to be only two types of notification status messages:

![oldnotifcations](https://user-images.githubusercontent.com/28930622/53669097-ea20bc00-3c43-11e9-953a-5cbb28e44559.png)


Now there are messages for Success, Timeout, and Fail (and the failure messages state WHY it didn't work):

![newnotifications](https://user-images.githubusercontent.com/28930622/53669113-002e7c80-3c44-11e9-82b1-d1d47b532179.png)

beeankha commented:
It appears that without the timeout field being an absolute requirement, this error gets thrown for what should be successful notifications:

![operationnowinprogress](https://user-images.githubusercontent.com/28930622/53744348-d4411000-3e6a-11e9-9f58-16c2190c3b2d.png)

I tried to filter for this error by putting `res.data.error != "[Errno 115] Operation now in progress"` in the  `awx/ui/client/src/notifications/notification-templates-list/list.controller.js` file, here: https://github.com/ansible/awx/blob/bf1769af6c6eae01a1740258e63c27c3324990a2/awx/ui/client/src/notifications/notification-templates-list/list.controller.js#L155

...but that just triggered the `Call to test notifications failed.` error found here:
https://github.com/ansible/awx/blob/bf1769af6c6eae01a1740258e63c27c3324990a2/awx/ui/client/src/notifications/notification-templates-list/list.controller.js#L167

I tinkered with that line but I couldn't seem to bypass that error.  


The only thing I've been able to do to make the timeout error work is to make the field a requirement, and have it be a range of 5 - 120 seconds.  The max timeout can be longer, but anything shorter than 5 seconds throws the `[Errno 115] Operation now in progress` error.  

![tooltip](https://user-images.githubusercontent.com/28930622/53759424-50008400-3e8e-11e9-9208-b2c0424e38ee.png)

beeankha commented:
I made the timeout optional, but when there's no timeout stated, this error shows up:

![CallToTestNotificationsFailed](https://user-images.githubusercontent.com/28930622/54150449-e5f85980-440e-11e9-972c-65aea4f1b263.png)

This is still related to that `res.data.error != "[Errno 115] Operation now in progress"` error being filtered out in `awx/awx/ui/client/src/notifications/notification-templates-list/list.controller.js`:

![OperationNowInProgress](https://user-images.githubusercontent.com/28930622/54150552-1e983300-440f-11e9-995f-32298852440e.png)

ryanpetrello commented:
One thing about this change (and the issue in general) that occurs to me is that I think a default of zero (no timeout) is actually a **really bad** default.

In my opinion, we should pick something sane, like 30 seconds, and let people change _that_ if they want.  I’d even argue **against** supporting “no timeout”, as it’s just going to result in hung processes if you've got your mail server misconfigured (bad).

```diff
diff --git a/awx/main/models/notifications.py b/awx/main/models/notifications.py
index 8bab75f0f..c0b4115be 100644
--- a/awx/main/models/notifications.py
+++ b/awx/main/models/notifications.py
@@ -187,7 +190,7 @@ class Notification(CreatedModifiedModel):
     )
     timeout = models.IntegerField(
         blank=True,
-        default=0,
+        default=30,
         help_text=_("The amount of time (in seconds) to run before the notification is canceled."),
     )
     body = JSONField(blank=True)
diff --git a/awx/ui/client/src/notifications/notificationTemplates.form.js b/awx/ui/client/src/notifications/notificationTemplates.form.js
index 3bc5b76d3..174d452cc 100644
--- a/awx/ui/client/src/notifications/notificationTemplates.form.js
+++ b/awx/ui/client/src/notifications/notificationTemplates.form.js
@@ -136,13 +136,13 @@ export default ['i18n', function(i18n) {
                 label: i18n._('Timeout'),
                 type: 'number',
                 integer: true,
-                min: 0,
-                default: 0,
+                min: 1,
+                default: 30,
                 spinner: true,
                 dataTitle: i18n._('Timeout'),
                 dataPlacement: 'right',
                 dataContainer: 'body',
-                awPopOver: "<p>" + i18n._("The amount of time (in seconds) before the email notification stops trying to reach the host and times out. Defaults to 0 for no time>
+                awPopOver: "<p>" + i18n._("The amount of time (in seconds) before the email notification stops trying to reach the host and times out.") + "</p>",
                 ngShow: "notification_type.value == 'email' ",
                 subForm: 'typeSubForm',
                 ngDisabled: '!(notification_template.summary_fields.user_capabilities.edit || canAdd)'
```

beeankha commented:
As of right now, the `username` and `password` fields throw an error when they're blank, which is the next issue that needs to be fixed.  

![usernamepasswordrequired](https://user-images.githubusercontent.com/28930622/54236112-393eda80-44e9-11e9-9d47-2ca03783f90d.png)

beeankha commented:
Everything seems happy now.

![EmailNotificationForm](https://user-images.githubusercontent.com/28930622/54290951-33e19e80-4582-11e9-991c-54e429db2a8b.png)

Timeout default is 30 seconds, ranges from 1 - 120 seconds.

![ToolTip](https://user-images.githubusercontent.com/28930622/54290976-3e039d00-4582-11e9-9a41-f263808901f8.png)

kialam reviewed (approved):
UI changes lgtm! Nice job.

ryanpetrello commented on awx/main/migrations/0064_notification_timeout.py:
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
```
On closer inspection, I don't think we actually _need_ this field (or migration) for the timeout to work.  We don't generally add backend-specific fields like this to the `main_notification` table unless all of the notifications use them (in the same way, you'll notice that `main_notification` doesn't have a username/password/host/port column).

The code that actually attempts to **send** the notification pulls the `timeout` (and other values) from the `main_notificationtemplate.notification_configuration` column:

```sql
awx-dev=> SELECT * FROM main_notificationtemplate;
-[ RECORD 1 ]--------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
id                         | 1
created                    | 2019-03-13 16:40:05.893517+00
modified                   | 2019-03-13 16:42:43.362269+00
description                |
name                       | Ryan Testing
notification_type          | email
notification_configuration | {"username": "user", "password": "$encrypted$UTF8$AESCBC$Z0FBQUFBQmNpVExLbFgyMm1VWEh2V1VOS3V6MTFQazgxYXU4a25hV21Pendud2RpNDk5VHN0bjBhUlFGZzQ3SHZRSWlCeEhkWHY5enVvQThtQnhIQS1LTXBGemdPT0NJU3c9PQ==", "host": "example.org", "recipients": ["joe@example.com"], "sender": "joe@example.com", "port": 25, "timeout": 3, "use_ssl": false, "use_tls": false}
created_by_id              | 1
modified_by_id             | 1
organization_id            | 1
```

https://github.com/ansible/awx/blob/devel/awx/main/models/notifications.py#L115-L128

With this diff:

```diff
diff --git a/awx/api/serializers.py b/awx/api/serializers.py
index 7dead78ea..ca454a0b1 100644
--- a/awx/api/serializers.py
+++ b/awx/api/serializers.py
@@ -4649,7 +4649,7 @@ class NotificationSerializer(BaseSerializer):
     class Meta:
         model = Notification
         fields = ('*', '-name', '-description', 'notification_template', 'error', 'status', 'notifications_sent',
-                  'notification_type', 'recipients', 'subject', 'timeout')
+                  'notification_type', 'recipients', 'subject')

     def get_related(self, obj):
         res = super(NotificationSerializer, self).get_related(obj)
diff --git a/awx/main/migrations/0064_notification_timeout.py b/awx/main/migrations/0064_notification_timeout.py
deleted file mode 100644
index 19f90eb04..000000000
--- a/awx/main/migrations/0064_notification_timeout.py
+++ /dev/null
@@ -1,20 +0,0 @@
-# -*- coding: utf-8 -*-
-# Generated by Django 1.11.16 on 2019-03-08 20:55
-from __future__ import unicode_literals
-
-from django.db import migrations, models
-
-
-class Migration(migrations.Migration):
-
-    dependencies = [
-        ('main', '0063_v350_org_host_limits'),
-    ]
-
-    operations = [
-        migrations.AddField(
-            model_name='notification',
-            name='timeout',
-            field=models.IntegerField(blank=True, default=0, help_text='The amount of time (in seconds) to run before the notification is canceled.'),
-        ),
-    ]
diff --git a/awx/main/migrations/0065_auto_20190312_1658.py b/awx/main/migrations/0065_auto_20190312_1658.py
deleted file mode 100644
index dd2edeae3..000000000
--- a/awx/main/migrations/0065_auto_20190312_1658.py
+++ /dev/null
@@ -1,20 +0,0 @@
-# -*- coding: utf-8 -*-
-# Generated by Django 1.11.16 on 2019-03-12 16:58
-from __future__ import unicode_literals
-
-from django.db import migrations, models
-
-
-class Migration(migrations.Migration):
-
-    dependencies = [
-        ('main', '0064_notification_timeout'),
-    ]
-
-    operations = [
-        migrations.AlterField(
-            model_name='notification',
-            name='timeout',
-            field=models.IntegerField(blank=True, default=30, help_text='The amount of time (in seconds) to run before the notification is canceled.'),
-        ),
-    ]
diff --git a/awx/main/models/notifications.py b/awx/main/models/notifications.py
index 6bd3e4c4e..39cea4917 100644
--- a/awx/main/models/notifications.py
+++ b/awx/main/models/notifications.py
@@ -185,11 +185,6 @@ class Notification(CreatedModifiedModel):
         default='',
         editable=False,
     )
-    timeout = models.IntegerField(
-        blank=True,
-        default=30,
-        help_text=_("The amount of time (in seconds) to run before the notification is canceled."),
-    )
     body = JSONField(blank=True)

     def get_absolute_url(self, request=None):
```

The timeout functionality seems to still work fine for me.

![image](https://user-images.githubusercontent.com/214912/54297572-883e4b80-458d-11e9-95eb-a8aa9724c4b3.png)

beeankha replied:
I got the same results, so this migration file has been removed.

ryanpetrello commented on awx/main/models/notifications.py:
```diff
@@ -185,6 +185,11 @@ class Meta:
         default='',
         editable=False,
     )
+    timeout = models.IntegerField(
```
Don't think we need this anymore, see big comment above.

beeankha replied:
Removed and tested, everything was fine so this requested change has been made!

ryanpetrello commented on awx/ui/client/src/notifications/notification-templates-list/list.controller.js:
```diff
@@ -131,7 +131,7 @@
                          ProcessErrors($scope, data, status, null, {
                              hdr: 'Error!',
-                             msg: 'Call to notifcatin templates failed. Notification returned status: ' + status
+                             msg: 'Call to notifcation templates failed. Notification returned status: ' + status
```
Still has a typo 😄

beeankha replied:
😂 
Fixing it!!

beeankha replied:
Fixed!

ryanpetrello commented on awx/ui/client/src/notifications/notification-templates-list/list.controller.js:
```diff
@@ -152,9 +152,14 @@
+                                     content: `<div><i class="fa fa-exclamation-triangle Toast-successIcon"></i> <b>${name}:</b> ${i18n._("Notification timed out.")}</div>`
+                                 });
+                                 $state.reload();
+                             } else if (res && res.data && res.data.status && res.data.status === "failed" && res.data.error !== "[Errno 115] Operation now in progress") {
```
Don't think we need this block anymore.

ryanpetrello replied:
Ah, sorry, I mean we don't need the `res.data.error !== "[Errno 115] Operation now in progress"` check.

beeankha replied:
Removed as requested, tested it and everything still works fine so ✔️

ryanpetrello commented on awx/api/serializers.py:
```diff
@@ -4649,7 +4649,7 @@ class NotificationSerializer(BaseSerializer):
         model = Notification
         fields = ('*', '-name', '-description', 'notification_template', 'error', 'status', 'notifications_sent',
-                  'notification_type', 'recipients', 'subject')
+                  'notification_type', 'recipients', 'subject', 'timeout')
```
Probably don't need this, see comment below.

beeankha replied:
👍

ryanpetrello commented:
@tvo318 when we merge this, the email notifications creation UI will have a new field, `timeout`, which takes an integer.  It will default to 30s:

![](https://user-images.githubusercontent.com/28930622/54290951-33e19e80-4582-11e9-991c-54e429db2a8b.png)

![](https://user-images.githubusercontent.com/28930622/54290976-3e039d00-4582-11e9-9a41-f263808901f8.png)

elyezer commented:
I was able to check the new timeout option and it works as expected. One thing though is that if I set the timeout for more than 55~60 the following alert is shown:

![Screenshot from 2019-03-14 14-37-04](https://user-images.githubusercontent.com/48132/54382629-c65b6e00-4666-11e9-8543-5fb61c821811.png)

So it seems that the call to test notification on the UI is timing our before the notification timeout and so the timeout alert:

![Screenshot from 2019-03-14 14-38-48](https://user-images.githubusercontent.com/48132/54382718-eb4fe100-4666-11e9-8221-ae4153882ec0.png)

Is never shown. With that said, I think the API is working fine but the UI still need to be updated.

elyezer commented:
I tested with recent changes @beeankha sent and now the timeout notification is displaying as it should. With that said I think this PR is ready to be merged.

beeankha commented:
recheck

beeankha commented:
recheck

ryanpetrello commented:
regate
